packages/drawing-response/src/drawing-response/drawable-main.jsx

```diff
@@ -11,16 +16,77 @@ import CircleDrawable from './drawable-circle';
 import EraserDrawable from './drawable-eraser';
 import Button from './button';
 
+const drawableToJSON = drawable => JSON.parse(JSON.stringify(drawable));
+
 class DrawableMain extends React.Component {
   constructor(props) {
     super(props);
     this.state = {
       drawables: [],
       newDrawable: [],
       textIsSelected: false,
+    };
+    this.debouncedSessionChange = debounce(this.handleSessionChange, 500);
+  }
+
+  UNSAFE_componentWillReceiveProps(nextProps) {
+    if (nextProps.session) {
+      const { TextEntry } = nextProps;
+      const { drawables: currentDrawables } = this.state;
+
+      if (nextProps.session.drawables) {
+        const newDrawables = cloneDeep(nextProps.session.drawables);
+        const drawablesString = JSON.stringify(currentDrawables);
+        const sessionDrawableString = JSON.stringify((newDrawables || []).map(drawable => omit(drawable, 'type')));
+
+        if (drawablesString !== sessionDrawableString) {
+          const drawableArray = nextProps.session.drawables;
+
+          this.setState({
+            drawables: drawableArray.map(drawable => this.getNewDrawableBasedOnType(drawable, drawable.type))
+          });
+        }
+
+        const currentTexts = TextEntry.all.map((text) => ({
+          ...text,
+          ...TextEntry[`text_${text.id}`] && TextEntry[`text_${text.id}`].attrs,
+          value: TextEntry[`textarea_${text.id}`] && TextEntry[`textarea_${text.id}`].value
+        }));
+
+        if (!isEqual(currentTexts, nextProps.session.texts)) {
+          TextEntry.setAll(nextProps.session.texts);
+        }
+      }
+
+    } else {
+      this.debouncedSessionChange();
     }
   }
 
+  handleSessionChange = () => {
+    const { onSessionChange, session, TextEntry } = this.props;
+    const { drawables } = this.state;
+
+    const newSession = {
+      drawables: drawables.map(drawable => ({
+        ...drawableToJSON(drawable),
```

Comment:
`drawables` it's an array of instances of certain classes that can render react components when the `render` function is called.
```
{drawables.map((drawable, key) => drawable.render({ ...drawableProps, key }))}
```
Is used below. Yeah it might be a good idea to change that.